gossip/src/duplicate_shred_handler.rs

```diff
@@ -0,0 +1,451 @@
+use {
+    crate::{
+        cluster_info_duplicate_shred_listener::ClusterInfoDuplicateShredHandler,
+        duplicate_shred::{DuplicateShred, Error},
+    },
+    itertools::Itertools,
+    log::*,
+    solana_ledger::{
+        blockstore::Blockstore, blockstore_meta::DuplicateSlotProof,
+        leader_schedule_cache::LeaderScheduleCache, shred::Shred,
+    },
+    solana_sdk::{clock::Slot, pubkey::Pubkey},
+    std::{
+        collections::{HashMap, HashSet},
+        sync::Arc,
+    },
+};
+
+const CLEANUP_EVERY_N_LOOPS: usize = 10;
```

Comment:
Why not use a `LruCache` instead?

Reply:
It's not exactly LruCache, we clean up when a slot is rooted so we don't care whether it's duplicate. LruCache is more aggressive, as non-rooted nodes could be long. Also I'm not sure LRU is the best policy here, because when there are multiple duplicate candidates, you might care more about finding old duplicates so the cluster could quickly converge on the same fork.